### PR TITLE
Mafia: support lynch modifiers

### PR DESCRIPTION
**Pull request by whalemer**
playerrole was requested by snaq

### CODE REVIEW

**Comment by HoeenCoder** [comment]
@whalemer conflicts/TS needs handling.

**Comment by whalemer** [comment]
i don't really like this implementation, ill rework it and rebase //eventually//

**Comment by HoeenCoder** [comment]
A suggestion for the future: add an `extraCount` property to lynch objects, when calculating hammer take this value into account.

eg: mayor votes player a
player a 4 -> 5
(player a extra lynches 0 -> 1)
hammer is 6, 5 + 1 = 6, hammer.

I feel this could really compress whats needed to make this work (I only glanced at the code here due to a lack of time)

**Comment by whalemer** [comment]
that's what lynch#trueCount is for, if i understand you correctly

**Comment by whalemer** [comment]
tfw you rewrite something from scratch and end up with 95% of the same code

updated

File: chat-plugins/mafia.js [hunk 1]
```diff
@@ -1516,13 +1629,14 @@ const pages = {
...
+				buf += `<p style="font-weight:bold;">${player.safeName} (${player.role ? player.getRole() : ''})${game.lynchModifiers[p] !== undefined ? ` (lynches are worth ${game.lynchModifiers[p]})` : ''}: <button class="button" name="send" value="/mafia kill ${room.id}, ${player.userid}">Kill</button> <button class="button" name="send" value="/mafia treestump ${room.id}, ${player.userid}">Make a Treestump (Kill)</button> <button class="button" name="send" value="/mafia spirit ${room.id}, ${player.userid}">Make a Restless Spirit (Kill)</button> <button class="button" name="send" value="/mafia spiritstump ${room.id}, ${player.userid}">Make a Restless Treestump (Kill)</button> <button class="button" name="send" value="/mafia sub ${room.id}, next, ${player.userid}">Force sub</button></p>`;
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: can you surround the buttons in <details> tags as per # suggestions on discord?

File: chat-plugins/mafia.js [hunk 2]
```diff
@@ -1516,13 +1629,14 @@ const pages = {
...
 			}
 			for (let d in game.dead) {
 				let dead = game.dead[d];
 				buf += `<p style="font-weight:bold;">${dead.safeName} (${dead.role ? dead.getRole() : ''})`;
 				if (dead.treestump) buf += ` (is a Treestump)`;
 				if (dead.restless) buf += ` (is a Restless Spirit)`;
+				if (game.lynchModifiers[d] !== undefined) buf += ` (lynches are worth ${game.lynchModifiers[d]})`;
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: same here with details

File: chat-plugins/mafia.js [hunk 3]
```diff
@@ -2594,7 +2838,27 @@ const commands = {
...
+		viewhostbans: function (target, room, user) {
+			if (!room || !room.mafiaEnabled) return this.errorReply(`Mafia is disabled for this room.`);
+			if (room.id !== 'mafia') return this.errorReply(`This command can only be used in the Mafia room.`);
+
+			if (!this.can('mute', null, room)) return false;
+			let bans = [];
+			for (const ban in hostBans) {
+				if (Date.now() > hostBans[ban]) {
+					delete hostBans[ban];
+					writeFile(BANS_FILE, hostBans);
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: write after looping so we only do it once.

File: chat-plugins/mafia.js [hunk 4]
```diff
@@ -2103,6 +2233,55 @@ const commands = {
...
+		love: function (target) {
+			this.parse(`/mafia applyhammermodifier ${target}, 1`);
+		},
+		hate: function (target) {
+			this.parse(`/mafia applyhammermodifier ${target}, -1`);
+		},
+		unhate: 'unlove',
+		unlove: function (target) {
+			this.parse(`/mafia applyhammermodifier ${target}, 0`);
+		},
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: these 3 could be combined into one command easily.

File: chat-plugins/mafia.js [hunk 5]
```diff
@@ -2103,6 +2233,55 @@ const commands = {
...
+		lynchless: 'voteless',
+		voteless: function (target) {
+			this.parse(`/mafia applylynchmodifier ${target}, 0`);
+		},
+		unvoteless: 'unmayor',
+		unlynchless: 'unmayor',
+		unmayor: function (target) {
+			this.parse(`/mafia applylynchmodifier ${target}, 1`);
+		},
```
Comment (HoeenCoder) [inline chat-plugins/mafia.js]: can also combine these 3

File: chat-plugins/mafia.js [hunk 6]
```diff
@@ -2104,6 +2271,71 @@ const commands = {
...
+			this.parse(`/mafia applyhammermodifier ${target}, ${mod}`);
+		},
+		doublevoter: 'mayor',
+		voteless: 'mayor',
+		unvoteless: 'mayor',
+		unmayor: 'mayor',
+		mayor: function (target, room, user, connection, cmd) {
+			let mod;
+			switch (cmd) {
+			case 'doublevoter': case 'mayor':
```
[the remark this answers is not in the document]
Comment (whalemer) [reply chat-plugins/mafia.js]: This is based off things like [editbattle aliases](https://github.com/Zarel/Pokemon-Showdown/blob/master/chat-commands.js#L3453), which to me are a similar enough usecase.

[the remark this answers is not in the document]
Comment (HoeenCoder) [reply chat-plugins/mafia.js]: Alright sure

**Pull request merged**